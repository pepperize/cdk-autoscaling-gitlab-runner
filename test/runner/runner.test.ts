import { App, Stack } from "aws-cdk-lib";
import { Capture, Template } from "aws-cdk-lib/assertions";
import { InstanceClass, InstanceSize, InstanceType, Vpc } from "aws-cdk-lib/aws-ec2";
import { ParameterTier, ParameterType, StringParameter } from "aws-cdk-lib/aws-ssm";
import { GitlabRunnerAutoscaling, GitlabRunnerAutoscalingJobRunnerProps } from "../../src";

describe("GitlabRunnerAutoscaling", () => {
  it("Should match snapshot when the runner is being used", () => {
    // Given
    const app = new App();
    const stack = new Stack(app, "MockStack", {
      env: {
        account: "0",
        region: "us-east-1",
      },
    });
    const vpc = Vpc.fromVpcAttributes(stack, "MyVpc", {
      vpcId: "vpc123",
      publicSubnetIds: ["pub1"],
      availabilityZones: ["us-east-1a"],
    });
    const token = new StringParameter(stack, "imported-token", {
      parameterName: "/gitlab-runner/token",
      stringValue: "auth-token",
      type: ParameterType.SECURE_STRING,
      tier: ParameterTier.STANDARD,
    });

    // When
    const runner = new GitlabRunnerAutoscaling(stack, "Runner", {
      concurrent: 10,
      checkInterval: 0,
      logFormat: "runner",
      logLevel: "info",
      network: {
        vpc: vpc,
      },
      runners: [
        {
          token: token,
          configuration: {
            name: "runner-one",
          },
        },
        {
          token: token,
          configuration: {
            name: "runner-two",
          },
        },
      ],
    });

    // Then
    const template = Template.fromStack(stack);
    expect(template).toMatchSnapshot();
    expect(runner.network.availabilityZone).toBe("us-east-1a");
  });

  it("Should have manager instance type set when it's set through props", () => {
    // Given
    const app = new App();
    const stack = new Stack(app, "MockStack", {
      env: {
        account: "0",
        region: "us-east-1",
      },
    });
    const vpc = Vpc.fromVpcAttributes(stack, "MyVpc", {
      vpcId: "vpc123",
      publicSubnetIds: ["pub1"],
      availabilityZones: ["us-east-1a"],
    });
    const token = new StringParameter(stack, "imported-token", {
      parameterName: "/gitlab-runner/token",
      stringValue: "auth-token",
      type: ParameterType.SECURE_STRING,
      tier: ParameterTier.STANDARD,
    });

    // When
    new GitlabRunnerAutoscaling(stack, "Runner", {
      network: {
        vpc: vpc,
      },
      runners: [
        {
          token: token,
          configuration: {
            name: "runner-one",
          },
        },
      ],
    });

    // Then
    const template = Template.fromStack(stack);
    const capture = new Capture();
    template.hasResourceProperties("AWS::AutoScaling::LaunchConfiguration", capture);
    expect(capture.asObject()).toMatchObject({ InstanceType: "t3.nano" });
  });

  it("Should have multiple runner configuration when it's set through props", () => {
    // Given
    const app = new App();
    const stack = new Stack(app, "MockStack", {
      env: {
        account: "0",
        region: "us-east-1",
      },
    });
    const token = new StringParameter(stack, "imported-token", {
      parameterName: "/gitlab-runner/token",
      stringValue: "auth-token",
      type: ParameterType.SECURE_STRING,
      tier: ParameterTier.STANDARD,
    });

    /**
     * t3.medium unprivileged work hours 10-18:00, 30 min idle time, teardown after 20 jobs, 1 hot standby
     */
    const runner1: GitlabRunnerAutoscalingJobRunnerProps = {
      token: token,
      configuration: {
        name: "gitlab-runner1",
        machine: {
          maxBuilds: 20, // teardown after 20 jobs
          idleCount: 0, // no instances except for the periods that are set at the autoscaling configuration
          autoscaling: [
            {
              idleCount: 1, // 1 hot standby
              idleTime: 1800, // 30 min idle time
              periods: ["* * 10-18 * * *"], // work hours 10-18:00
            },
          ],
        },
        docker: {
          privileged: false, // unprivileged
        },
      },
      instanceType: InstanceType.of(InstanceClass.T3, InstanceSize.MEDIUM), // t3.medium
    };

    /**
     * t3.small unprivileged, work hours 13-17:00, 15 min idle time, teardown after 5 jobs
     */
    const runner2: GitlabRunnerAutoscalingJobRunnerProps = {
      token: token,
      configuration: {
        name: "gitlab-runner2",
        machine: {
          maxBuilds: 5, // teardown after 5 jobs
          idleCount: 0, // no instances except for the periods that are set at the autoscaling configuration
          autoscaling: [
            {
              periods: ["* * 13-17 * * *"], // work hours 13-17:00
              idleCount: 1, // 1 hot standby ?
              idleTime: 900, // 15 min idle time
            },
          ],
        },
        docker: {
          privileged: false, // unprivileged
        },
      },
      instanceType: InstanceType.of(InstanceClass.T3, InstanceSize.SMALL), // t3.medium
    };

    /**
     * t3.xlarge, privileged, no hot standby, teardown after 1 job
     */
    const runner3: GitlabRunnerAutoscalingJobRunnerProps = {
      token: token,
      configuration: {
        name: "gitlab-runner3",
        machine: {
          maxBuilds: 1, // teardown after 1 job
          idleCount: 0, // no hot standby
          autoscaling: [
            {
              idleCount: 0, // no hot standby
              periods: ["* * * * * *"], // always use this configuration
            },
          ],
        },
        // privileged by default
      },
      instanceType: InstanceType.of(InstanceClass.T3, InstanceSize.XLARGE), // t3.xlarge
    };

    // When
    new GitlabRunnerAutoscaling(stack, "Runner", {
      runners: [runner1, runner2, runner3],
    });

    // Then
    const template = Template.fromStack(stack);
    expect(template).toMatchSnapshot();
    template.hasResourceProperties("AWS::AutoScaling::LaunchConfiguration", {});
    expect(JSON.stringify(template)).toMatch("gitlab-runner1");
    expect(JSON.stringify(template)).toMatch("gitlab-runner2");
    expect(JSON.stringify(template)).toMatch("gitlab-runner3");
  });

  it("Should generate a unique gitlab runner name when provided a configuration without the name set", () => {
    // Given
    const app = new App();
    const stack = new Stack(app, "MockStack", {
      env: {
        account: "0",
        region: "us-east-1",
      },
    });

    const token = new StringParameter(stack, "imported-token", {
      parameterName: "/gitlab-runner/token",
      stringValue: "auth-token",
      type: ParameterType.SECURE_STRING,
      tier: ParameterTier.STANDARD,
    });

    // When
    new GitlabRunnerAutoscaling(stack, "Runner", {
      runners: [
        {
          token: token,
          configuration: {},
        },
      ],
    });

    // Then
    const template = Template.fromStack(stack);
    template.hasResourceProperties("AWS::AutoScaling::LaunchConfiguration", {});
  });
});
