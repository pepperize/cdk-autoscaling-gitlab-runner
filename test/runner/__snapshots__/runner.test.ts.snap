// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GitlabRunnerAutoscaling Should match snapshot 1`] = `
Object {
  "Parameters": Object {
    "AssetParameters15684a15d07860e99d2a8079150ad33dd2cb743677fcd7016dd07345e1b69538ArtifactHash154EE95B": Object {
      "Description": "Artifact hash for asset \\"15684a15d07860e99d2a8079150ad33dd2cb743677fcd7016dd07345e1b69538\\"",
      "Type": "String",
    },
    "AssetParameters15684a15d07860e99d2a8079150ad33dd2cb743677fcd7016dd07345e1b69538S3BucketB021608E": Object {
      "Description": "S3 bucket for asset \\"15684a15d07860e99d2a8079150ad33dd2cb743677fcd7016dd07345e1b69538\\"",
      "Type": "String",
    },
    "AssetParameters15684a15d07860e99d2a8079150ad33dd2cb743677fcd7016dd07345e1b69538S3VersionKeyC73B105A": Object {
      "Description": "S3 key for asset version \\"15684a15d07860e99d2a8079150ad33dd2cb743677fcd7016dd07345e1b69538\\"",
      "Type": "String",
    },
    "SsmParameterValueawsserviceamiamazonlinuxlatestamzn2amihvmx8664ebsC96584B6F00A464EAD1953AFF4B05118Parameter": Object {
      "Default": "/aws/service/ami-amazon-linux-latest/amzn2-ami-hvm-x86_64-ebs",
      "Type": "AWS::SSM::Parameter::Value<AWS::EC2::Image::Id>",
    },
  },
  "Resources": Object {
    "CacheBucket41D9D0B0": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "BucketEncryption": Object {
          "ServerSideEncryptionConfiguration": Array [
            Object {
              "ServerSideEncryptionByDefault": Object {
                "SSEAlgorithm": "aws:kms",
              },
            },
          ],
        },
        "BucketName": "mockstack-runner-cache-0-us-east-1",
        "LifecycleConfiguration": Object {
          "Rules": Array [
            Object {
              "ExpirationInDays": 30,
              "Status": "Enabled",
            },
          ],
        },
        "PublicAccessBlockConfiguration": Object {
          "BlockPublicAcls": true,
          "BlockPublicPolicy": true,
          "IgnorePublicAcls": true,
          "RestrictPublicBuckets": true,
        },
        "Tags": Array [
          Object {
            "Key": "aws-cdk:auto-delete-objects",
            "Value": "true",
          },
        ],
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Delete",
    },
    "CacheBucketAutoDeleteObjectsCustomResource12E72BBC": Object {
      "DeletionPolicy": "Delete",
      "DependsOn": Array [
        "CacheBucketPolicyE2F00ED4",
      ],
      "Properties": Object {
        "BucketName": Object {
          "Ref": "CacheBucket41D9D0B0",
        },
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "CustomS3AutoDeleteObjectsCustomResourceProviderHandler9D90184F",
            "Arn",
          ],
        },
      },
      "Type": "Custom::S3AutoDeleteObjects",
      "UpdateReplacePolicy": "Delete",
    },
    "CacheBucketPolicyE2F00ED4": Object {
      "Properties": Object {
        "Bucket": Object {
          "Ref": "CacheBucket41D9D0B0",
        },
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
              ],
              "Effect": "Allow",
              "Principal": Object {
                "AWS": Object {
                  "Fn::GetAtt": Array [
                    "CustomS3AutoDeleteObjectsCustomResourceProviderRole3B1BD092",
                    "Arn",
                  ],
                },
              },
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "CacheBucket41D9D0B0",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "CacheBucket41D9D0B0",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
    "CustomS3AutoDeleteObjectsCustomResourceProviderHandler9D90184F": Object {
      "DependsOn": Array [
        "CustomS3AutoDeleteObjectsCustomResourceProviderRole3B1BD092",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters15684a15d07860e99d2a8079150ad33dd2cb743677fcd7016dd07345e1b69538S3BucketB021608E",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters15684a15d07860e99d2a8079150ad33dd2cb743677fcd7016dd07345e1b69538S3VersionKeyC73B105A",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters15684a15d07860e99d2a8079150ad33dd2cb743677fcd7016dd07345e1b69538S3VersionKeyC73B105A",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Description": Object {
          "Fn::Join": Array [
            "",
            Array [
              "Lambda function for auto-deleting objects in ",
              Object {
                "Ref": "CacheBucket41D9D0B0",
              },
              " S3 bucket.",
            ],
          ],
        },
        "Handler": "__entrypoint__.handler",
        "MemorySize": 128,
        "Role": Object {
          "Fn::GetAtt": Array [
            "CustomS3AutoDeleteObjectsCustomResourceProviderRole3B1BD092",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "CustomS3AutoDeleteObjectsCustomResourceProviderRole3B1BD092": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Sub": "arn:\${AWS::Partition}:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "ManagerAutoscalingGroupASGE4841860": Object {
      "CreationPolicy": Object {
        "ResourceSignal": Object {
          "Count": 1,
          "Timeout": "PT15M",
        },
      },
      "Metadata": Object {
        "AWS::CloudFormation::Init": Object {
          "config": Object {
            "files": Object {
              "/etc/gitlab-runner/config.toml": Object {
                "content": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "concurrent = 10
check_interval = 0
log_format = \\"runner\\"
log_level = \\"info\\"

[[runners]]
name = \\"gitlab-runner\\"
url = \\"https://gitlab.com\\"
limit = 10
output_limit = 52_428_800
executor = \\"docker+machine\\"
environment = [ \\"DOCKER_DRIVER=overlay2\\", \\"DOCKER_TLS_CERTDIR=/certs\\" ]
token = \\"\\"

  [runners.docker]
  tls_verify = false
  image = \\"docker:19.03.5\\"
  privileged = true
  cap_add = [ \\"CAP_SYS_ADMIN\\" ]
  wait_for_services_timeout = 300
  disable_cache = false
  volumes = [ \\"/certs/client\\", \\"/cache\\" ]
  shm_size = 0

  [runners.machine]
  IdleCount = 0
  IdleTime = 300
  MaxBuilds = 20
  MachineDriver = \\"amazonec2\\"
  MachineName = \\"gitlab-runner-%s\\"
  MachineOptions = [
  \\"amazonec2-request-spot-instance=true\\",
  \\"amazonec2-spot-price=0.03\\",
  \\"amazonec2-instance-type=t3.micro\\",
  \\"amazonec2-ami=ami-1234\\",
  \\"amazonec2-region=us-east-1\\",
  \\"amazonec2-vpc-id=vpc123\\",
  \\"amazonec2-zone=a\\",
  \\"amazonec2-subnet-id=pub1\\",
  \\"amazonec2-security-group=MockStack-RunnersSecurityGroup\\",
  \\"amazonec2-use-private-address=true\\",
  \\"amazonec2-iam-instance-profile=",
                      Object {
                        "Ref": "RunnersInstanceProfile",
                      },
                      "\\"
]

    [[runners.machine.autoscaling]]
    Periods = [ \\"* * 7-22 * * mon-fri *\\" ]
    IdleCount = 1
    IdleTime = 1_800
    Timezone = \\"Etc/UTC\\"

  [runners.cache]
  Type = \\"s3\\"
  Shared = true

    [runners.cache.s3]
    ServerAddress = \\"s3.",
                      Object {
                        "Ref": "AWS::URLSuffix",
                      },
                      "\\"
    BucketName = \\"",
                      Object {
                        "Ref": "CacheBucket41D9D0B0",
                      },
                      "\\"
    BucketLocation = \\"us-east-1\\"
",
                    ],
                  ],
                },
                "encoding": "plain",
                "group": "gitlab-runner",
                "mode": "000600",
                "owner": "gitlab-runner",
              },
              "/etc/rsyslog.d/25-gitlab-runner.conf": Object {
                "content": ":programname, isequal, \\"gitlab-runner\\" /var/log/gitlab-runner.log",
                "encoding": "plain",
                "group": "root",
                "mode": "000644",
                "owner": "root",
              },
            },
            "services": Object {
              "sysvinit": Object {
                "gitlab-runner": Object {
                  "enabled": true,
                  "ensureRunning": true,
                  "files": Array [
                    "/etc/gitlab-runner/config.toml",
                  ],
                },
                "rsyslog": Object {
                  "enabled": true,
                  "ensureRunning": true,
                  "files": Array [
                    "/etc/rsyslog.d/25-gitlab-runner.conf",
                  ],
                },
              },
            },
          },
          "configSets": Object {
            "default": Array [
              "repositories",
              "packages",
              "config",
              "restart",
            ],
          },
          "packages": Object {
            "commands": Object {
              "10-docker-machine": Object {
                "command": "curl -L https://gitlab-docker-machine-downloads.s3.amazonaws.com/v0.16.2-gitlab.12/docker-machine-\`uname -s\`-\`uname -m\` > /tmp/docker-machine && install /tmp/docker-machine /usr/bin/docker-machine",
              },
              "20-gitlab-runner-start": Object {
                "command": "gitlab-runner start",
              },
            },
            "packages": Object {
              "yum": Object {
                "docker": Array [],
                "gitlab-runner": Array [],
                "tzdata": Array [],
              },
            },
          },
          "repositories": Object {
            "commands": Object {
              "10-gitlab-runner": Object {
                "command": "curl -L https://packages.gitlab.com/install/repositories/runner/gitlab-runner/script.rpm.sh | bash",
              },
            },
          },
          "restart": Object {
            "commands": Object {
              "10-gitlab-runner-restart": Object {
                "command": "gitlab-runner restart",
              },
            },
          },
        },
      },
      "Properties": Object {
        "DesiredCapacity": "1",
        "LaunchConfigurationName": Object {
          "Ref": "ManagerAutoscalingGroupLaunchConfig078E7BD8",
        },
        "MaxSize": "1",
        "MinSize": "1",
        "Tags": Array [
          Object {
            "Key": "Name",
            "PropagateAtLaunch": true,
            "Value": "MockStack/ManagerAutoscalingGroup",
          },
        ],
        "VPCZoneIdentifier": Array [
          "pub1",
        ],
      },
      "Type": "AWS::AutoScaling::AutoScalingGroup",
      "UpdatePolicy": Object {
        "AutoScalingRollingUpdate": Object {
          "PauseTime": "PT15M",
          "SuspendProcesses": Array [
            "HealthCheck",
            "ReplaceUnhealthy",
            "AZRebalance",
            "AlarmNotification",
            "ScheduledActions",
          ],
          "WaitOnResourceSignals": true,
        },
        "AutoScalingScheduledAction": Object {
          "IgnoreUnmodifiedGroupSizeProperties": true,
        },
      },
    },
    "ManagerAutoscalingGroupInstanceProfile7C061EF0": Object {
      "Properties": Object {
        "Roles": Array [
          Object {
            "Ref": "ManagerRole08FB5B64",
          },
        ],
      },
      "Type": "AWS::IAM::InstanceProfile",
    },
    "ManagerAutoscalingGroupLaunchConfig078E7BD8": Object {
      "DependsOn": Array [
        "ManagerRoleDefaultPolicy9C31350E",
        "ManagerRole08FB5B64",
      ],
      "Properties": Object {
        "IamInstanceProfile": Object {
          "Ref": "ManagerAutoscalingGroupInstanceProfile7C061EF0",
        },
        "ImageId": Object {
          "Ref": "SsmParameterValueawsserviceamiamazonlinuxlatestamzn2amihvmx8664ebsC96584B6F00A464EAD1953AFF4B05118Parameter",
        },
        "InstanceType": "t3.nano",
        "SecurityGroups": Array [
          Object {
            "Fn::GetAtt": Array [
              "ManagerSecurityGroup288FD437",
              "GroupId",
            ],
          },
        ],
        "UserData": Object {
          "Fn::Base64": Object {
            "Fn::Join": Array [
              "",
              Array [
                "#!/bin/bash
yum update -y aws-cfn-bootstrap
# fingerprint: a2ec99bc087bc298
(
  set +e
  /opt/aws/bin/cfn-init -v --region ",
                Object {
                  "Ref": "AWS::Region",
                },
                " --stack ",
                Object {
                  "Ref": "AWS::StackName",
                },
                " --resource ManagerAutoscalingGroupASGE4841860 -c default
  /opt/aws/bin/cfn-signal -e $? --region ",
                Object {
                  "Ref": "AWS::Region",
                },
                " --stack ",
                Object {
                  "Ref": "AWS::StackName",
                },
                " --resource ManagerAutoscalingGroupASGE4841860
  cat /var/log/cfn-init.log >&2
)",
              ],
            ],
          },
        },
      },
      "Type": "AWS::AutoScaling::LaunchConfiguration",
    },
    "ManagerRole08FB5B64": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "ec2.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          "arn:aws:iam::aws:policy/service-role/AmazonEC2RoleforSSM",
        ],
        "Policies": Array [
          Object {
            "PolicyDocument": Object {
              "Statement": Array [
                Object {
                  "Action": Array [
                    "s3:ListObjects*",
                    "s3:GetObject*",
                    "s3:DeleteObject*",
                    "s3:PutObject*",
                  ],
                  "Effect": "Allow",
                  "Resource": Object {
                    "Fn::Join": Array [
                      "",
                      Array [
                        Object {
                          "Fn::GetAtt": Array [
                            "CacheBucket41D9D0B0",
                            "Arn",
                          ],
                        },
                        "/*",
                      ],
                    ],
                  },
                },
                Object {
                  "Action": "s3:ListBucket",
                  "Effect": "Allow",
                  "Resource": Object {
                    "Fn::GetAtt": Array [
                      "CacheBucket41D9D0B0",
                      "Arn",
                    ],
                  },
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "Cache",
          },
          Object {
            "PolicyDocument": Object {
              "Statement": Array [
                Object {
                  "Action": Array [
                    "ec2:CreateKeyPair",
                    "ec2:DeleteKeyPair",
                    "ec2:ImportKeyPair",
                    "ec2:Describe*",
                  ],
                  "Effect": "Allow",
                  "Resource": "*",
                },
                Object {
                  "Action": Array [
                    "ec2:CreateTags",
                    "ssm:UpdateInstanceInformation",
                  ],
                  "Condition": Object {
                    "ForAllValues:StringEquals": Object {
                      "aws:TagKeys": Array [
                        "Name",
                      ],
                    },
                    "StringLike": Object {
                      "aws:RequestTag/Name": "*gitlab-runner-*",
                    },
                  },
                  "Effect": "Allow",
                  "Resource": "*",
                },
                Object {
                  "Action": Array [
                    "ec2:RequestSpotInstances",
                    "ec2:CancelSpotInstanceRequests",
                  ],
                  "Condition": Object {
                    "ArnEqualsIfExists": Object {
                      "ec2:Vpc": Object {
                        "Fn::Join": Array [
                          "",
                          Array [
                            "arn:",
                            Object {
                              "Ref": "AWS::Partition",
                            },
                            ":ec2:us-east-1:0:vpc/vpc123",
                          ],
                        ],
                      },
                    },
                    "StringEqualsIfExists": Object {
                      "ec2:Region": "us-east-1",
                    },
                  },
                  "Effect": "Allow",
                  "Resource": "*",
                },
                Object {
                  "Action": "ec2:RunInstances",
                  "Condition": Object {
                    "StringEquals": Object {
                      "ec2:InstanceProfile": Object {
                        "Ref": "RunnersInstanceProfile",
                      },
                      "ec2:InstanceType": Array [
                        "t3.micro",
                      ],
                    },
                  },
                  "Effect": "Allow",
                  "Resource": "*",
                },
                Object {
                  "Action": Array [
                    "ec2:TerminateInstances",
                    "ec2:StopInstances",
                    "ec2:StartInstances",
                    "ec2:RebootInstances",
                  ],
                  "Condition": Object {
                    "StringLike": Object {
                      "ec2:ResourceTag/Name": "*gitlab-runner-*",
                    },
                  },
                  "Effect": "Allow",
                  "Resource": "*",
                },
                Object {
                  "Action": "iam:PassRole",
                  "Effect": "Allow",
                  "Resource": Object {
                    "Fn::GetAtt": Array [
                      "RunnersRole1AC39F4E",
                      "Arn",
                    ],
                  },
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "Runners",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "ManagerRoleDefaultPolicy9C31350E": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "cloudformation:DescribeStackResource",
                "cloudformation:SignalResource",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Ref": "AWS::StackId",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "ManagerRoleDefaultPolicy9C31350E",
        "Roles": Array [
          Object {
            "Ref": "ManagerRole08FB5B64",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "ManagerSecurityGroup288FD437": Object {
      "Properties": Object {
        "GroupDescription": "Security group for GitLab Runners Manager.",
        "SecurityGroupEgress": Array [
          Object {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow all outbound traffic by default",
            "IpProtocol": "-1",
          },
        ],
        "VpcId": "vpc123",
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "RunnersInstanceProfile": Object {
      "Properties": Object {
        "Roles": Array [
          Object {
            "Ref": "RunnersRole1AC39F4E",
          },
        ],
      },
      "Type": "AWS::IAM::InstanceProfile",
    },
    "RunnersRole1AC39F4E": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "ec2.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          "arn:aws:iam::aws:policy/service-role/AmazonEC2RoleforSSM",
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "RunnersSecurityGroup2A22C282": Object {
      "Properties": Object {
        "GroupDescription": "Security group for GitLab Runners.",
        "GroupName": "MockStack-RunnersSecurityGroup",
        "SecurityGroupEgress": Array [
          Object {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow all outbound traffic by default",
            "IpProtocol": "-1",
          },
        ],
        "VpcId": "vpc123",
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "RunnersSecurityGroupfromMockStackManagerSecurityGroup7EE94F53228E94ADCB": Object {
      "Properties": Object {
        "Description": "SSH traffic from Manager",
        "FromPort": 22,
        "GroupId": Object {
          "Fn::GetAtt": Array [
            "RunnersSecurityGroup2A22C282",
            "GroupId",
          ],
        },
        "IpProtocol": "tcp",
        "SourceSecurityGroupId": Object {
          "Fn::GetAtt": Array [
            "ManagerSecurityGroup288FD437",
            "GroupId",
          ],
        },
        "ToPort": 22,
      },
      "Type": "AWS::EC2::SecurityGroupIngress",
    },
    "RunnersSecurityGroupfromMockStackManagerSecurityGroup7EE94F5323764EC94E30": Object {
      "Properties": Object {
        "Description": "SSH traffic from Docker",
        "FromPort": 2376,
        "GroupId": Object {
          "Fn::GetAtt": Array [
            "RunnersSecurityGroup2A22C282",
            "GroupId",
          ],
        },
        "IpProtocol": "tcp",
        "SourceSecurityGroupId": Object {
          "Fn::GetAtt": Array [
            "ManagerSecurityGroup288FD437",
            "GroupId",
          ],
        },
        "ToPort": 2376,
      },
      "Type": "AWS::EC2::SecurityGroupIngress",
    },
  },
}
`;
